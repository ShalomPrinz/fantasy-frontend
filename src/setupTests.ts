// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import axios from "axios";
import { setupServer } from "msw/node";

import { handlers } from "../test";
import * as TeamContext from "./contexts/TeamContext";
import * as UserContext from "./features/authentication/contexts/UserContext";
import { LeagueInfo, Message, Player, Team, User } from "./types";

jest.setTimeout(10000);

axios.defaults.baseURL = "";

const server = setupServer(...handlers);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

interface MockedUserContext {
  logout?: () => Promise<void>;
  state?: UserContext.UserState;
  user?: User;
}

interface MockedUser {
  inbox?: Message[];
  leagues?: LeagueInfo[];
  name?: string;
  players?: Player[];
}

export const getTestUser = (user?: MockedUser): User => ({
  id: "id",
  inbox: user?.inbox || [],
  leagues: user?.leagues || [],
  name: user?.name || "",
  team: new Team(user?.players || []),
});

export const mockUser = (context: MockedUserContext) => {
  const userContext = jest.spyOn(UserContext, "useUser");

  const { logout, state, user } = context;
  const mockedLogout = typeof logout !== "undefined" ? logout : async () => {};
  const mockedState =
    typeof state !== "undefined" ? state : UserContext.UserState.LOGGED_USER;
  const mockedUser = typeof user !== "undefined" ? user : getTestUser();

  userContext.mockImplementation(() => ({
    loading: false,
    login: async () => {},
    logout: mockedLogout,
    register: async () => {},
    state: mockedState,
    user: mockedUser,
    invalidate: async () => {},
  }));

  return userContext.mockRestore;
};

export const mockTeam = (team: Team) => {
  const teamContext = jest.spyOn(TeamContext, "useTeamState");
  teamContext.mockImplementation(() => team);
};

export * from "../test";
