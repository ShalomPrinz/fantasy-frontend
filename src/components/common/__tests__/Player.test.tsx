import { hoverElement, render, screen, unhoverElement } from "setupTests";

import { Player } from "../..";

describe("Player", () => {
  it("should render Player component", () => {
    const { asFragment } = render(
      <Player name="Shalom" team="Dortmund" width={10} widthUnits="px" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe("Hover", () => {
    it("should show full player name on hover", async () => {
      const { asFragment, user } = render(
        <Player
          name="Robert Lewandowski"
          team="Dortmund"
          width={10}
          widthUnits="px"
        />
      );

      const firstRender = asFragment();
      const element = screen.getByText(/Robert/i);
      await hoverElement(user, element);

      expect(firstRender).not.toEqual(asFragment());
      expect(asFragment()).toMatchSnapshot();
    });

    it("should get back to normal when hover stops", async () => {
      const { asFragment, user } = render(
        <Player
          name="Robert Lewandowski"
          team="Dortmund"
          width={10}
          widthUnits="px"
        />
      );

      const firstRender = asFragment();
      const element = screen.getByText(/Robert/i);

      await hoverElement(user, element);
      expect(firstRender).not.toEqual(asFragment());

      await unhoverElement(user, element);
      expect(firstRender).toEqual(asFragment());
    });
  });
});
