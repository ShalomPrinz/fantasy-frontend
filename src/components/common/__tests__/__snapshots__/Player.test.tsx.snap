// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player Hover should show full player name on hover 1`] = `
<DocumentFragment>
  <div
    class="text-center"
    style="width: 14px;"
  >
    <img
      class=""
      src="Dortmund.png"
      style="width: 10px;"
      title="Dortmund"
    />
    <div
      aria-describedby="tooltip-top"
      class="text-truncate bg-default rounded p-1 w-100"
    >
      Robert Lewandowski
    </div>
  </div>
</DocumentFragment>
`;

exports[`Player should render Player component 1`] = `
<DocumentFragment>
  <div
    class="text-center"
    style="width: 14px;"
  >
    <img
      class=""
      src="Dortmund.png"
      style="width: 10px;"
      title="Dortmund"
    />
    <div
      class="text-truncate bg-default rounded p-1 w-100"
    >
      Shalom
    </div>
  </div>
</DocumentFragment>
`;
